### PR TITLE
Arbitrary Decodable

### PR DESCRIPTION
**Pull request by stephencelis**
Fun one to explore!

``` swift
struct User: Decodable {
  let id: Int
  let name: String
}

dump(
  User.arbitrary
)
// ▿ User
//   - id: -145456
//   - name: "迭㐴湔᤻"
```

I had to drag out a recent snapshot that cleaned up the `Codable` interfaces quite a bit, but we could probably back-port it if we wanna use this before the next beta.

### CODE REVIEW

File: Sources/LCG/LCG.swift [hunk 1]
```diff
@@ -11,12 +11,18 @@ public struct Seed {
...
   }
   
-  public static func random() -> IO<Seed> {
-    return IO { Seed(Int(arc4random_uniform(UInt32(N)))) }
+  public static let random = IO {
+    (Seed.init(seed:) <<< Int.init <<< arc4random_uniform <<< UInt32.init) <| N
+  }
+}
+
+extension Seed: ExpressibleByIntegerLiteral {
```
Comment (stephencelis) [inline Sources/LCG/LCG.swift:23]: Seemed a good use-case for `ExpressibleByIntegerLiteral`. We don't wanna pass `Seed`s around by accident to `Int` receivers, but we do wanna make it easy to re-run a seed explicitly.

File: Sources/QuickCheck/ArbitraryDecodable.swift [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+private let genInt32 = genInt.map(Int32.init)
+private let genInt64 = genInt.map(Int64.init)
+private let genUInt = choose(0..<2_000_000).map(UInt.init)
+private let genUInt8 = choose(0..<Int(UInt8.max)).map(UInt8.init)
+private let genUInt16 = choose(0..<Int(UInt16.max)).map(UInt16.init)
+private let genUInt32 = choose(0..<2_000_000).map(UInt32.init)
+private let genUInt64 = choose(0..<2_000_000).map(UInt64.init)
+private let genFloat = genDouble.map(Float.init)
+
+private final class ArbitraryDecoder: Decoder {
```
Comment (stephencelis) [inline Sources/QuickCheck/ArbitraryDecodable.swift]: We we could make this a `struct` (under the hood it handles the mutation for us), all the sub-containers message back to increment the seed. I'll explore in another commit, but it was feeling messy.

File: Sources/QuickCheck/ArbitraryDecodable.swift [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+      self.codingPath = decoder.codingPath
+      self.decoder = decoder
+    }
+
+    func contains(_ key: Key) -> Bool {
+      return true
+    }
+
+    func decodeNil(forKey key: Key) throws -> Bool {
+      return self.decoder.run(uniform.map { $0 >= 0.75 })
```
Comment (stephencelis) [inline Sources/QuickCheck/ArbitraryDecodable.swift:61]: Decoders don't decode `Optional` as a type, so I'm recreating the logic behind our `optionalOf` generator here.

File: Sources/QuickCheck/ArbitraryDecodable.swift [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    }
+
+    func decode<T>(_ type: T.Type, forKey key: Key) throws -> T where T: Decodable {
+      self.decoder.codingPath.append(key)
```
Comment (stephencelis) [inline Sources/QuickCheck/ArbitraryDecodable.swift:65]: This `codingPath` tracking is mainly for making sure the `fatalError`s provide more context.

File: Sources/QuickCheck/ArbitraryDecodable.swift [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+  struct KeyedContainer<Key: CodingKey>: KeyedDecodingContainerProtocol {
+    let allKeys: [Key]
+    let codingPath: [CodingKey]
+    let decoder: ArbitraryDecoder
+
+    init(decoder: ArbitraryDecoder) {
+      let genKeys = arrayOf(genInt.map(Key.init(intValue:))) |> scale { $0 * 2 }
+
+      self.allKeys = decoder.run(genKeys).flatMap { $0 }
```
Comment (stephencelis) [inline Sources/QuickCheck/ArbitraryDecodable.swift]: Decoders don't decode `Dictionary` as a type, so I'm using the `arrayOf` generator to build up a random collection of keys.

File: Sources/QuickCheck/ArbitraryDecodable.swift [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+    let codingPath: [CodingKey]
+    let count: Int?
+    var currentIndex = 0
+    let decoder: ArbitraryDecoder
+    var isAtEnd: Bool { return currentIndex == count }
+
+    init(decoder: ArbitraryDecoder) {
+      self.codingPath = decoder.codingPath
+      print(decoder.genState.size)
+      self.count = decoder.run(choose(0..<decoder.genState.size) |> suchThat <| { $0 % 2 == 0 })
```
Comment (stephencelis) [inline Sources/QuickCheck/ArbitraryDecodable.swift]: Decoders don't decode `Array` as a type, so I'm using the logic behind the `arrayOf` generator to build up a random count. The `suchThat <| { $0 % 2 == 0 }` is a workaround for collections like `Set` which require even keys—hoping there's a better way.

File: Sources/QuickCheck/ArbitraryDecodable.swift [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+      self.codingPath = decoder.codingPath
+      self.decoder = decoder
+    }
+
+    func decodeNil() -> Bool {
+      return self.decoder.run(uniform.map { $0 >= 0.75 })
+    }
+
+    func decode(_ type: Bool.Type) throws -> Bool {
+      return self.decoder.run(genBool)
```
Comment (stephencelis) [inline Sources/QuickCheck/ArbitraryDecodable.swift:144]: The latest Swift snapshot lets us define these base type decodings in one place! When I tried on beta 3 I had to reproduce these in each other decoding container, twice! :(

File: Sources/QuickCheck/Gen.swift [hunk 8]
```diff
@@ -45,6 +45,12 @@ public func resize<A>(_ size: Int) -> (Gen<A>) -> Gen<A> {
   }
 }
 
+public func scale<A>(_ f: @escaping (Int) -> Int) -> (Gen<A>) -> Gen<A> {
```
Comment (stephencelis) [inline Sources/QuickCheck/Gen.swift:49]: I'd used this in the Haskell version. Not sure why the PureScript version doesn't define it!

**Comment by stephencelis** [comment]
(Tests are gonna be failing till the next beta or till we conform to the older `Codable` protocols.)

**Review by mbrandonw — APPROVED** [review]
wonderful! it compiles and tests pass on newest Xcode+Swift, but unfortunately I wasnt quite sure how to resolve the merge conflicts with master, so will leave this with you!

**Pull request merged**